File: world/chunk.py
import numpy as np
from core.config import Config
from world.block_types import BLOCKS


class Chunk:
    """Represents a 16x256x16 chunk of blocks"""
    
    def __init__(self, x, z):
        self.x = x  # Chunk X coordinate
        self.z = z  # Chunk Z coordinate
        self.blocks = np.zeros(
            (Config.CHUNK_SIZE, Config.CHUNK_HEIGHT, Config.CHUNK_SIZE),
            dtype=np.uint8
        )
        self.is_dirty = True  # Needs mesh regeneration
        self.is_generated = False
        self.mesh_data = None
        self.neighbors = {
            'north': None,
            'south': None,
            'east': None,
            'west': None
        }
    
    def get_block(self, x, y, z):
        """Get block at local coordinates"""
        if 0 <= x < Config.CHUNK_SIZE and 0 <= y < Config.CHUNK_HEIGHT and 0 <= z < Config.CHUNK_SIZE:
            return self.blocks[x, y, z]
        return 0  # Air for out of bounds
    
    def set_block(self, x, y, z, block_id):
        """Set block at local coordinates"""
        if 0 <= x < Config.CHUNK_SIZE and 0 <= y < Config.CHUNK_HEIGHT and 0 <= z < Config.CHUNK_SIZE:
            if self.blocks[x, y, z] != block_id:
                self.blocks[x, y, z] = block_id
                self.is_dirty = True
                
                # Mark neighboring chunks dirty if block is on edge
                if x == 0 and self.neighbors['west']:
                    self.neighbors['west'].is_dirty = True
                elif x == Config.CHUNK_SIZE - 1 and self.neighbors['east']:
                    self.neighbors['east'].is_dirty = True
                
                if z == 0 and self.neighbors['north']:
                    self.neighbors['north'].is_dirty = True
                elif z == Config.CHUNK_SIZE - 1 and self.neighbors['south']:
                    self.neighbors['south'].is_dirty = True
    
    def is_block_visible(self, x, y, z):
        """Check if a block face should be rendered"""
        block_id = self.get_block(x, y, z)
        if block_id == 0:  # Air blocks not visible
            return False
        
        # Check all 6 faces
        faces_visible = 0
        
        # Check each neighbor
        for dx, dy, dz in [(1,0,0), (-1,0,0), (0,1,0), (0,-1,0), (0,0,1), (0,0,-1)]:
            nx, ny, nz = x + dx, y + dy, z + dz
            
            # Get neighbor block (might be from adjacent chunk)
            if 0 <= nx < Config.CHUNK_SIZE and 0 <= ny < Config.CHUNK_HEIGHT and 0 <= nz < Config.CHUNK_SIZE:
                neighbor_id = self.blocks[nx, ny, nz]
            else:
                # Edge of chunk - check neighbor chunk
                neighbor_id = self._get_neighbor_block(nx, ny, nz)
            
            # Face is visible if neighbor is air or transparent
            if neighbor_id == 0 or BLOCKS.is_transparent(neighbor_id):
                faces_visible += 1
        
        return faces_visible > 0
    
    def _get_neighbor_block(self, x, y, z):
        """Get block from neighboring chunk if on edge"""
        if y < 0 or y >= Config.CHUNK_HEIGHT:
            return 0  # Air above/below world
        
        chunk = None
        local_x, local_z = x, z
        
        if x < 0:
            chunk = self.neighbors['west']
            local_x = Config.CHUNK_SIZE - 1
        elif x >= Config.CHUNK_SIZE:
            chunk = self.neighbors['east']
            local_x = 0
        
        if z < 0:
            chunk = self.neighbors['north']
            local_z = Config.CHUNK_SIZE - 1
        elif z >= Config.CHUNK_SIZE:
            chunk = self.neighbors['south']
            local_z = 0
        
        if chunk and chunk.is_generated:
            return chunk.get_block(local_x, y, local_z)
        
        return 0  # Air for unloaded chunks
    
    def get_world_position(self):
        """Get world position of chunk origin"""
        return (
            self.x * Config.CHUNK_SIZE,
            0,
            self.z * Config.CHUNK_SIZE
        )
    
    def __repr__(self):
        return f"Chunk({self.x}, {self.z})"

File: src/chunk.c
#include "chunk.h"
#include <stdlib.h>
#include <string.h>

Chunk* chunk_create(int x, int z) {
    Chunk* chunk = (Chunk*)malloc(sizeof(Chunk));
    if (!chunk) return NULL;
    
    chunk->x = x;
    chunk->z = z;
    chunk->is_generated = false;
    chunk->is_dirty = true;
    chunk->mesh = NULL;
    
    // Initialize neighbors to NULL
    chunk->north = NULL;
    chunk->south = NULL;
    chunk->east = NULL;
    chunk->west = NULL;
    
    // Clear blocks
    memset(chunk->blocks, 0, sizeof(chunk->blocks));
    
    return chunk;
}

void chunk_destroy(Chunk* chunk) {
    if (chunk) {
        // Mesh is freed by renderer
        free(chunk);
    }
}

BlockType chunk_get_block(Chunk* chunk, int x, int y, int z) {
    if (!chunk) return BLOCK_AIR;
    
    if (x >= 0 && x < CHUNK_SIZE && 
        y >= 0 && y < CHUNK_HEIGHT && 
        z >= 0 && z < CHUNK_SIZE) {
        return chunk->blocks[x][y][z];
    }
    
    return BLOCK_AIR;
}

void chunk_set_block(Chunk* chunk, int x, int y, int z, BlockType type) {
    if (!chunk) return;
    
    if (x >= 0 && x < CHUNK_SIZE && 
        y >= 0 && y < CHUNK_HEIGHT && 
        z >= 0 && z < CHUNK_SIZE) {
        
        if (chunk->blocks[x][y][z] != type) {
            chunk->blocks[x][y][z] = type;
            chunk->is_dirty = true;
            
            // Mark neighboring chunks dirty if on edge
            if (x == 0 && chunk->west) {
                chunk->west->is_dirty = true;
            } else if (x == CHUNK_SIZE - 1 && chunk->east) {
                chunk->east->is_dirty = true;
            }
            
            if (z == 0 && chunk->north) {
                chunk->north->is_dirty = true;
            } else if (z == CHUNK_SIZE - 1 && chunk->south) {
                chunk->south->is_dirty = true;
            }
        }
    }
}

BlockType chunk_get_neighbor_block(Chunk* chunk, int x, int y, int z) {
    if (!chunk) return BLOCK_AIR;
    
    // Check if out of vertical bounds
    if (y < 0 || y >= CHUNK_HEIGHT) {
        return BLOCK_AIR;
    }
    
    // If within this chunk, return directly
    if (x >= 0 && x < CHUNK_SIZE && z >= 0 && z < CHUNK_SIZE) {
        return chunk->blocks[x][y][z];
    }
    
    // Check neighboring chunks
    Chunk* neighbor = NULL;
    int local_x = x;
    int local_z = z;
    
    if (x < 0) {
        neighbor = chunk->west;
        local_x = CHUNK_SIZE - 1;
    } else if (x >= CHUNK_SIZE) {
        neighbor = chunk->east;
        local_x = 0;
    }
    
    if (z < 0) {
        neighbor = chunk->north;
        local_z = CHUNK_SIZE - 1;
    } else if (z >= CHUNK_SIZE) {
        neighbor = chunk->south;
        local_z = 0;
    }
    
    if (neighbor && neighbor->is_generated) {
        return chunk_get_block(neighbor, local_x, y, local_z);
    }
    
    return BLOCK_AIR;
}

bool chunk_is_block_visible(Chunk* chunk, int x, int y, int z) {
    BlockType block = chunk_get_block(chunk, x, y, z);
    
    if (block == BLOCK_AIR) {
        return false;
    }
    
    // Check all 6 faces
    int dx[] = {1, -1, 0, 0, 0, 0};
    int dy[] = {0, 0, 1, -1, 0, 0};
    int dz[] = {0, 0, 0, 0, 1, -1};
    
    for (int i = 0; i < 6; i++) {
        int nx = x + dx[i];
        int ny = y + dy[i];
        int nz = z + dz[i];
        
        BlockType neighbor = chunk_get_neighbor_block(chunk, nx, ny, nz);
        
        // Face is visible if neighbor is air or transparent
        if (neighbor == BLOCK_AIR || block_is_transparent(neighbor)) {
            return true;
        }
    }
    
    return false;
}